### PR TITLE
F45 add jasmine tests for idea view model reduced

### PR DESCRIPTION
**Pull request by gmcgibbon**
Jasmine unit tests for our idea KO CoffeeScript classes.

### CODE REVIEW

File: spec/javascripts/views/ideas_view_model_spec.js.coffee [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+        votes: 512
+        project_id: 7331
+      }
+    ]
+    @subject = new WinniVote.IdeasViewModel(@ideas)
+
+  describe "#constructor", ->
+    it "has ideas",         ->
+      expect(
+        {
```
Comment (dbernar1) [inline spec/javascripts/views/ideas_view_model_spec.js.coffee]: Hey Gannon - would there be any use in the code for a IdeasViewModel.ideas_json() which would encapsulate this?

**Comment by amirci** [comment]
For the IdeaViewModel spec I think that we can consider _IdeaToggle_ as an external dependency and do something like:

``` Coffee
    @toggler = jasmine.createSpyObj('Toggler', ['toggle'])
    # Stub the constructor
    spyOn(WinniVote, 'IdeaToggle').andReturn @toggler
```

That way then we just need to assert (without doing a _spyOn_):

``` Coffee
expect(@toggler.toggle).toHaveBeenCalled()
```

File: spec/javascripts/views/ideas_view_model_spec.js.coffee (hunk 1 above)
Comment (gmcgibbon) [inline spec/javascripts/views/ideas_view_model_spec.js.coffee]: @dbernar1 
Do you mean encapsulating the creation of mock data? We don't really have access to a factory to create the mock data. I could try abstracting it out to a separate file and class for data definition at least.

**Comment by gmcgibbon** [comment]
This is similar to what I was already doing. What are the benefits to spying on IdeaToggle directly as opposed to the internal toggler of an IdeaViewModel? 

``` coffeescript
# before
spyOn(@subject.toggler, "toggle")

# assert
expect(@subject.toggler.toggle).toHaveBeenCalled()
```

**Comment by amirci** [comment]
The difference is that you mock the entire class thus reducing the external dependency to that class.

This is similar to the idea of dependency injection that we discussed.

**Comment by gmcgibbon** [comment]
Good point, I will make this change and look at separating the mock data definition from unit tests.

**Comment by amirci** [comment]
I'm not sure why you call the factory data JSON.... 

JSON is a string representation (text based) and we are using Javascript objects.

Let's imagine you rename the method to _createMany_ the code can use the _list comprehension_ to create all the ideas, something like:

``` coffee
@createMany: -> (@create i  for i in [1 .. noOfIdeas])
```

**Comment by gmcgibbon** [comment]
I named the factory method getIdeasJSON because it isn't returning an actual Idea object. This way if we wanted to make another method that instantiates the JSON data to an Idea object automatically the different method signatures would make sense. 

``` coffeescript
# the difference between returning data for an Idea
@ideaObj = new WinniVote.IdeaViewModel(WinniVote.IdeaFactory.getJSONIdea(22))
# and returning a fully instantiated Idea object
@ideaObj = WinniVote.IdeaFactory.getIdea(22)
```

The window.ideas variable defined in our ideas index view is defined as an array of JS objects. Is it just how this array was created that makes it acceptable to call it's contents JSON? 

If so, the factory method I created is supposed to mimic the creation of idea JS objects _from_ JSON right? In my opinion, calling these factory methods something else just needlessly complicates things. 

Regardless, I will change it if you still disagree. I'm not too sure about naming it **createMany** because the content of the return is vague. So, if I can't name it **getJSONIdea**, how does **getIdeaData** sound?

**Comment by gmcgibbon** [comment]
``` coffeescript
it "has ideas", ->
      expect(ko.mapping.toJS(i) for i in @subject.ideas()).toEqual(@ideasJSON)
```

The unmapping code seems to work great, nice find! I can create a helper to further encapsulate this process but it is a one liner as is. What do you think?

**Comment by dbernar1** [comment]
Great, I'm glad that worked out, Gannon.

As far as a helper, my rule of thumb is: if you can think of another way that you think you might like better, try it out, and then decide whether to keep it or throw it away. ( Though I really like helpers these days, as you can see in something like this: https://gist.github.com/dbernar1/10285679#file-3-language-support-coffee )

Have a good day!

**Comment by amirci** [comment]
First of all I think we went too far, creating a Factory wasn't part of the feature. Also as part of refactoring I would have aimed to do the simplest implementation possible.

Again JSON is a text format and not sure why we are considering it. Convert to JSON and un-map seems to deviate a bit of our goal and solves a problem we don't have. Why have a factory that returns JSON? 

Perhaps we should leave it the way it was and think better about JS factories and research about them and then if we need to, let's discuss an implementation.

A comment on the factory: I don't see the confusion of the naming of the methods of _IdeaFactory_. Similar to **FactoryGirl** if we use _create_ and _createMany_ or _createList_ would have the same meaning. Where is the ambiguity?

``` Coffee
# When calling to create one
IdeaFactory.create()

# or calling createList to create 3 ideas
IdeaFactory.createList 3
```

**Comment by swalberg** [comment]
I'm not against the idea of introducing some simple factories at this point just to begin exploration. I probably would have stayed with the static data to start, but I like that you're trying things and leaving some breadcrumbs for the next guy. This is a learning project, after all.

One problem I see with the current implementation has to do with the naming, both of the factory methods and the use of JSON when we're talking about a plain old javascript object. The methods for `FooFactory` shouldn't have `foo` in them -- I know I'm getting a `Foo` because I asked `FooFactory`! Similarly by convention we know we're only really ever going to ask for our ViewModel objects and a simple hash, so it's not necessary to put that in the method name.

I think what you're trying to do is implement the `attributes_for` that we see in FactoryGirl so that you can test the constructor. So while I agree with what Amir is saying in that `IdeaFactory.create()` should give me an `Idea` and not an intermediate representation, I think that renaming your methods to something like `IdeaFactory.attributes` and `IdeaFactory.attributesArray` might better communicate your intent. And then `IdeaFactory.create` becomes `new WinniVote.IdeasViewModel(Winnivote.IdeaFactory.attributes())`. (but you don't need that yet, so don't implement it!)

And to Amir's other point, we should be thinking of how we want to continue with Factories after this.  We don't want to go too far down the path when we're not sure what we're trying to solve. This exploration is cool, but let's not get to the point where we're spending a lot of time refactoring tests to undo previous decisions.

**Comment by amirci** [comment]
I think you missed the part when Sean says "you don't need that now" :)

I'm going to take the liberty to modify the factory to a very simple version following the convention of factory girl.

Also I am going to remove the helpers because first we don't need them now, and second they are not more clear in my opinion that just using a standard construct like list comprehension or _map_.

I'll merge it after that

**Pull request closed without merging**